table/builder.go

```diff
@@ -133,38 +126,87 @@ func (b *Builder) addHelper(key []byte, v y.ValueStruct) {
 	h := header{
 		plen: uint16(len(key) - len(diffKey)),
 		klen: uint16(len(diffKey)),
-		vlen: uint16(v.EncodedSize()),
-		prev: b.prevOffset, // prevOffset is the location of the last key-value added.
+		vlen: uint32(v.EncodedSize()),
 	}
-	b.prevOffset = uint32(b.buf.Len()) - b.baseOffset // Remember current offset for the next Add call.
+
+	// store current entry's offset
+	b.entryOffsets = append(b.entryOffsets, uint32(b.buf.Len())-b.baseOffset)
 
 	// Layout: header, diffKey, value.
-	var hbuf [10]byte
+	var hbuf [8]byte
 	h.Encode(hbuf[:])
 	b.buf.Write(hbuf[:])
 	b.buf.Write(diffKey) // We only need to store the key difference.
 
 	v.EncodeTo(b.buf)
-	b.counter++ // Increment number of keys added for this current block.
 }
 
-func (b *Builder) finishBlock() {
-	// When we are at the end of the block and Valid=false, and the user wants to do a Prev,
-	// we need a dummy header to tell us the offset of the previous key-value pair.
-	b.addHelper([]byte{}, y.ValueStruct{})
+func (b *Builder) finishBlock() error {
+	idx := &pb.BlockIndex{
```

Comment:
Will it be useful to add a comment with the layout that is used in this function?